packages/hmr-plugin/src/internal/state-event-loop.ts

```diff
@@ -0,0 +1,43 @@
+import { NgxsBootstrapper, StateContext, StateStream } from '@ngxs/store';
+import { NgModuleRef } from '@angular/core';
+import { Subscription } from 'rxjs';
+
+import { getStateContext } from './state-context';
+import { CallStackFrame, NgxsHmrLifeCycle } from '../symbols';
+import { getBootstrap, getStateFromHmrStorage, getStateStream } from './common';
+
+export function stateEventLoop<T extends NgxsHmrLifeCycle<S>, S>(
+  ref: NgModuleRef<T>,
+  frame: CallStackFrame<S>
+) {
+  const stateContext: StateContext<S> | undefined = getStateContext<T, S>(ref);
+  if (stateContext) {
+    const previousState: Partial<S> = getStateFromHmrStorage<S>();
+    const existSavedState: boolean = Object.keys(previousState).length > 0;
+
+    // get event subscriptions
+    const stateStream: StateStream | null = getStateStream<T>(ref);
+    const bootstrap: NgxsBootstrapper | null = getBootstrap(ref);
+    const canBeUpdateState: boolean = Boolean(existSavedState && stateStream && bootstrap);
+
+    if (canBeUpdateState) {
+      let eventId: number;
+      const waitTime = 100;
+
+      bootstrap!.appBootstrapped$.subscribe(() => {
+        const stateStreamSub: Subscription = stateStream!.subscribe(() => {
+          // setTimeout used for zone detection after set hmr state
+          clearInterval(eventId);
+          // only need to get the last message
+          eventId = window.setTimeout(() => {
```

Comment:
Do we really need to use `setTimeout`?
Also, no need to write `window.`.

Reply:
Yes, because we need triggers for NgZone